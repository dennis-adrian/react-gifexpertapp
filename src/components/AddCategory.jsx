import React, { useState } from 'react';
import PropTypes from 'prop-types';

const AddCategory = ({ setCategories }) => {

    const [inputValue, setInputValue] = useState('')

    const handleInputChange = (e) => {
        setInputValue(e.target.value);

    }

    const handleSubmit = (e) => {
        //prevenir el comportamiento por defecto del form
        //es decir, prevenir que se recarge
        e.preventDefault();


        if (inputValue.trim().length > 2) {

            //categs representa el valor inicial
            //el valor del arreglo categories
            //que se encuentra en el componoente GifExpertApp
            //y a ese valor le agregamos el input value
            setCategories(categs => [inputValue, ...categs]);
            setInputValue('');
        }
    }

    return (
        <form onSubmit={handleSubmit}>

            <h1>{inputValue}</h1>
            <input
                type="text"
                value={inputValue}
                onChange={handleInputChange}
            />
        </form>
    )
};

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
export default AddCategory;