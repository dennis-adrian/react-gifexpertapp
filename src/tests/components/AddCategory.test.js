import React from 'react';
import '@testing-library/jest-dom';

import { shallow } from 'enzyme';
import AddCategory from '../../components/AddCategory';

describe('Testing on <AddCategory />', () => {

    const setCategories = jest.fn();
    let wrapper = shallow(<AddCategory setCategories={setCategories} />);

    beforeEach(() => {
        //se llama para limpiar todas las simulaciones que tengamos de algo
        jest.clearAllMocks();
        wrapper = shallow(<AddCategory setCategories={setCategories} />);
    })


    test('should render correctly', () => {

        expect(wrapper).toMatchSnapshot();
    })

    test('should change the input box', () => {
        const input = wrapper.find('input');

        //simulamos el e.target.value que se necesita en nuestro componente
        const value = 'hello world'
        //el objeto representa el evento (e)
        input.simulate('change', {
            target: { value }
        });

    })

    test('should not post the infomation on submit', () => {
        wrapper.find('form').simulate('submit', {
            preventDefault() { }
        })

        expect(setCategories).not.toHaveBeenCalled();
    })

    test('should call setCategories and clean the input textbox', () => {
        const value = 'hello world';

        //simlando el cambio en el input
        wrapper.find('input').simulate('change', {
            target: { value }
        });
        //simulando el submit
        wrapper.find('form').simulate('submit', {
            preventDefault() { }
        });
        //verificamos que se haya llamado a la funci贸n
        expect(setCategories).toHaveBeenCalled();
        /*podemos aseguranos que la funci贸n haya sido
        llamada exactamente una vez*/
        expect(setCategories).toHaveBeenCalledTimes(1);
        /** Podemos asegurarnos que se haya llamado con una
         * funci贸n y no con cualquier otro tipo de dato
        */
        expect(setCategories).toHaveBeenCalledWith(expect.any(Function));



        //verificar si el input se vaci贸
        expect(wrapper.find('input').prop('value')).toBe('');

    })

})