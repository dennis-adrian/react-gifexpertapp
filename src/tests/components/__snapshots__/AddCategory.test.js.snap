// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing on <AddCategory /> should render correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <h1 />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
